In [11]:
import pandas as pd
import numpy as np

In [12]:
pd.set_option('display.max_rows', None, 'display.max_columns', None, 'display.max_colwidth', -1)

C:\Users\sdn11\Anaconda\lib\site-packages\ipykernel_launcher.py:1: FutureWarning: Passing a negative integer is deprecated in version 1.0 and will not be supported in future version. Instead, use None to not limit the column width.
  """Entry point for launching an IPython kernel.


## 데이터 불러오기

In [13]:
df_raw = pd.read_csv("data/몽데이크_Open.csv", encoding="utf-8")
print(df_raw.shape)
df_raw.head()

(5364, 9)


qplay_question_id  \
0  3504                
1  3675                
2  3779                
3  3649                
4  3388                

                                                                                                          text  \
0  연립부등식 `{(-x-3 >= -2x+1), (2x+a <= 10) :}`의 해가 존재하도록 하는 `a`의 최댓값을 구하면?                                         
1  부등식 `[x]^2-2[x]-3 < 0`의 해는?                                                                                   
2  점 `(1, 2)`를 지나며 중심이 직선 `x-y-1=0` 위에 있고 `x`축에 접하는 원의 방정식은?                                                     
3  `x`, `y`가 정수일 때, 방정식 `xy+2x-3y=7`을 만족하는 순서쌍 `(x, y)`의 개수는?                                                    
4  평행이동 `(x, y)->(x+2, y-7)`에 의해 포물선 `y=x^2`을 평행이동하였더니 포물선 `y=x^2+px+q`가 되었다고 할 때, 상수 `p`, `q`에 대하여 `p+q`의 값은?   

         qtid                      name subtitle  question_difficulty  \
0  H1S1-08-05  해를 갖거나 갖지 않는 연립일차부등식      NaN      3                     
1  H1S1-08-11  절댓값 또는 가우스 기호가 포함된 이차부등식  NaN      3                     
2  H1S1-11-05  x축 또는 y축에 접하는 원의 방정식      NaN      3                     
3  H1S1-07-18  정수 및 자연수 조건을 갖는 부정방정식     NaN      3                     
4  H1S1-12-03  도형의 평행이동 (2)              원        2                     

   question_type_difficulty  grade purpose  
0  2.0                       10     open    
1  2.0                       10     open    
2  2.0                       10     open    
3  3.0                       10     open    
4  2.0                       10     open

In [24]:
len(df_raw['qtid'].unique())

592

In [25]:
len(df_raw['name'].unique())

592

In [23]:
sorted(df_raw['qtid'].unique())

['H1S1-01-01',
 'H1S1-01-02',
 'H1S1-01-03',
 'H1S1-01-04',
 'H1S1-01-05',
 'H1S1-01-06',
 'H1S1-01-07',
 'H1S1-01-08',
 'H1S1-01-09',
 'H1S1-02-01',
 'H1S1-02-02',
 'H1S1-02-03',
 'H1S1-02-04',
 'H1S1-02-05',
 'H1S1-02-06',
 'H1S1-02-07',
 'H1S1-02-08',
 'H1S1-02-09',
 'H1S1-02-10',
 'H1S1-02-11',
 'H1S1-02-12',
 'H1S1-02-13',
 'H1S1-02-14',
 'H1S1-03-01',
 'H1S1-03-02',
 'H1S1-03-03',
 'H1S1-03-04',
 'H1S1-03-05',
 'H1S1-03-06',
 'H1S1-03-07',
 'H1S1-03-08',
 'H1S1-03-09',
 'H1S1-03-10',
 'H1S1-03-11',
 'H1S1-04-01',
 'H1S1-04-02',
 'H1S1-04-03',
 'H1S1-04-04',
 'H1S1-04-05',
 'H1S1-04-06',
 'H1S1-04-07',
 'H1S1-04-08',
 'H1S1-04-09',
 'H1S1-05-01',
 'H1S1-05-02',
 'H1S1-05-03',
 'H1S1-05-04',
 'H1S1-05-05',
 'H1S1-05-06',
 'H1S1-05-07',
 'H1S1-05-08',
 'H1S1-05-09',
 'H1S1-05-10',
 'H1S1-05-11',
 'H1S1-05-12',
 'H1S1-05-13',
 'H1S1-05-14',
 'H1S1-05-15',
 'H1S1-06-01',
 'H1S1-06-02',
 'H1S1-06-03',
 'H1S1-06-04',
 'H1S1-06-05',
 'H1S1-06-06',
 'H1S1-06-07',
 'H1S1-06-08',
 'H1S1-06-

In [29]:
df_raw[df_raw['qtid']=='H1S1-05-07']['name'].unique()

array(['이차식이 완전제곱식이 되는 조건'], dtype=object)

In [20]:
# grouped = abalone['whole_weight'].groupby(abalone['sex'])

df_raw['name'].groupby(df_raw['qtid'])

In [5]:
df_splited_qtid = df_raw["qtid"].str.split("-", expand=True,)
df_splited_qtid[:5]

0   1   2
0  H1S1  08  05
1  H1S1  08  11
2  H1S1  11  05
3  H1S1  07  18
4  H1S1  12  03

In [6]:
# Split qtid column into 학년, 중단원, 소단원 three independent columns
df_wrangled = df_raw
df_wrangled[["학년", "중단원", "소단원"]] = df_splited_qtid
df_wrangled = df_wrangled.drop(columns=["qtid", "purpose", "grade"])
df_wrangled.head()

qplay_question_id  \
0  3504                
1  3675                
2  3779                
3  3649                
4  3388                

                                                                                                          text  \
0  연립부등식 `{(-x-3 >= -2x+1), (2x+a <= 10) :}`의 해가 존재하도록 하는 `a`의 최댓값을 구하면?                                         
1  부등식 `[x]^2-2[x]-3 < 0`의 해는?                                                                                   
2  점 `(1, 2)`를 지나며 중심이 직선 `x-y-1=0` 위에 있고 `x`축에 접하는 원의 방정식은?                                                     
3  `x`, `y`가 정수일 때, 방정식 `xy+2x-3y=7`을 만족하는 순서쌍 `(x, y)`의 개수는?                                                    
4  평행이동 `(x, y)->(x+2, y-7)`에 의해 포물선 `y=x^2`을 평행이동하였더니 포물선 `y=x^2+px+q`가 되었다고 할 때, 상수 `p`, `q`에 대하여 `p+q`의 값은?   

                       name subtitle  question_difficulty  \
0  해를 갖거나 갖지 않는 연립일차부등식      NaN      3                     
1  절댓값 또는 가우스 기호가 포함된 이차부등식  NaN      3                     
2  x축 또는 y축에 접하는 원의 방정식      NaN      3                     
3  정수 및 자연수 조건을 갖는 부정방정식     NaN      3                     
4  도형의 평행이동 (2)              원        2                     

   question_type_difficulty    학년 중단원 소단원  
0  2.0                       H1S1  08  05  
1  2.0                       H1S1  08  11  
2  2.0                       H1S1  11  05  
3  3.0                       H1S1  07  18  
4  2.0                       H1S1  12  03

In [7]:
df_preprocessed = pd.read_csv("data/data_preprocessed.csv", encoding="utf-8")
print(df_preprocessed.shape)
df_preprocessed.head()

(5364, 16)


chapter        qtid  \
0  H1S1-08  H1S1-08-05   
1  H1S1-08  H1S1-08-11   
2  H1S1-11  H1S1-11-05   
3  H1S1-07  H1S1-07-18   
4  H1S1-12  H1S1-12-03   

                                                                                                          text  \
0  연립부등식 `{(-x-3 >= -2x+1), (2x+a <= 10) :}`의 해가 존재하도록 하는 `a`의 최댓값을 구하면?                                         
1  부등식 `[x]^2-2[x]-3 < 0`의 해는?                                                                                   
2  점 `(1, 2)`를 지나며 중심이 직선 `x-y-1=0` 위에 있고 `x`축에 접하는 원의 방정식은?                                                     
3  `x`, `y`가 정수일 때, 방정식 `xy+2x-3y=7`을 만족하는 순서쌍 `(x, y)`의 개수는?                                                    
4  평행이동 `(x, y)->(x+2, y-7)`에 의해 포물선 `y=x^2`을 평행이동하였더니 포물선 `y=x^2+px+q`가 되었다고 할 때, 상수 `p`, `q`에 대하여 `p+q`의 값은?   

                                                       math_field  \
0  ['{(-x-3 >= -2x+1), (2x+a <= 10) :}', 'a']                       
1  ['[x]^2-2[x]-3 < 0']                                             
2  ['(1, 2)', 'x-y-1=0', 'x']                                       
3  ['x', 'y', 'xy+2x-3y=7', '(x, y)']                               
4  ['(x, y)->(x+2, y-7)', 'y=x^2', 'y=x^2+px+q', 'p', 'q', 'p+q']   

                                                                hangul_field  \
0  ['연립부등식', '해', '존재하도록', '하', '최댓값', '구하']                                   
1  ['부등식', '해']                                                                
2  ['점', '지나', '중심', '직선', '위', '있', '축', '접하', '원', '방정식']                    
3  ['정수일', '때', '방정식', '만족하', '순서쌍', '개수']                                     
4  ['평행이동', '의해', '포물선', '평행이동하였더', '포물선', '되었다', '할', '때', '상수', '대하', '값']   

  tag_field                      name subtitle  qplay_question_id  \
0  []        해를 갖거나 갖지 않는 연립일차부등식      NaN      3504                
1  []        절댓값 또는 가우스 기호가 포함된 이차부등식  NaN      3675                
2  []        x축 또는 y축에 접하는 원의 방정식      NaN      3779                
3  []        정수 및 자연수 조건을 갖는 부정방정식     NaN      3649                
4  []        도형의 평행이동 (2)              원        3388                

   question_difficulty  question_type_difficulty  grade purpose  \
0  3                    2.0                       10     open     
1  3                    2.0                       10     open     
2  3                    2.0                       10     open     
3  3                    3.0                       10     open     
4  2                    2.0                       10     open     

               word2vec_pred_chapter  \
0  ['H1S1-08', 'H1S1-08', 'H1S1-08']   
1  ['H1S1-08', 'HSU1-06', 'H1S1-08']   
2  ['H1S1-11', 'H1S1-12', 'H1S1-11']   
3  ['H1S1-07', 'H1S1-10', 'H1S1-08']   
4  ['H1S1-12', 'H1S1-06', 'H1S2-03']   

                                                           word2vec_pred_name  \
0  ['연립이차부등식', '정수인 해의 개수가 주어진 연립부등식', '해를 갖거나 갖지 않는 연립일차부등식']                  
1  ['이차부등식', '삼각부등식 (2)', '절댓값 또는 가우스 기호가 포함된 이차부등식']                           
2  ['x축 또는 y축에 접하는 원의 방정식', '평행이동과 대칭이동의 활용', 'x축과 y축에 동시에 접하는 원의 방정식']         
3  ['정수 및 자연수 조건을 갖는 부정방정식', '평행한 두 직선 사이의 거리', '이차부등식을 이용하여 식의 최댓값·최솟값 구하기']   
4  ['도형의 평행이동 (2)', '식의 최댓값과 최솟값 (2)', '코시-슈바르츠의 부등식 (1)']                      

  tfidf_pred_chapter  
0  H1S1-08            
1  H1S1-08            
2  H1S1-11            
3  H1S1-07            
4  H1S1-01

In [8]:
# Split qtid column into 학년, 중단원, 소단원 three independent columns
df_preprocessed[["학년", "중단원", "소단원"]] = df_splited_qtid
df_preprocessed = df_preprocessed.drop(columns=["qtid", "purpose", "grade"])
df_preprocessed.head()

chapter  \
0  H1S1-08   
1  H1S1-08   
2  H1S1-11   
3  H1S1-07   
4  H1S1-12   

                                                                                                          text  \
0  연립부등식 `{(-x-3 >= -2x+1), (2x+a <= 10) :}`의 해가 존재하도록 하는 `a`의 최댓값을 구하면?                                         
1  부등식 `[x]^2-2[x]-3 < 0`의 해는?                                                                                   
2  점 `(1, 2)`를 지나며 중심이 직선 `x-y-1=0` 위에 있고 `x`축에 접하는 원의 방정식은?                                                     
3  `x`, `y`가 정수일 때, 방정식 `xy+2x-3y=7`을 만족하는 순서쌍 `(x, y)`의 개수는?                                                    
4  평행이동 `(x, y)->(x+2, y-7)`에 의해 포물선 `y=x^2`을 평행이동하였더니 포물선 `y=x^2+px+q`가 되었다고 할 때, 상수 `p`, `q`에 대하여 `p+q`의 값은?   

                                                       math_field  \
0  ['{(-x-3 >= -2x+1), (2x+a <= 10) :}', 'a']                       
1  ['[x]^2-2[x]-3 < 0']                                             
2  ['(1, 2)', 'x-y-1=0', 'x']                                       
3  ['x', 'y', 'xy+2x-3y=7', '(x, y)']                               
4  ['(x, y)->(x+2, y-7)', 'y=x^2', 'y=x^2+px+q', 'p', 'q', 'p+q']   

                                                                hangul_field  \
0  ['연립부등식', '해', '존재하도록', '하', '최댓값', '구하']                                   
1  ['부등식', '해']                                                                
2  ['점', '지나', '중심', '직선', '위', '있', '축', '접하', '원', '방정식']                    
3  ['정수일', '때', '방정식', '만족하', '순서쌍', '개수']                                     
4  ['평행이동', '의해', '포물선', '평행이동하였더', '포물선', '되었다', '할', '때', '상수', '대하', '값']   

  tag_field                      name subtitle  qplay_question_id  \
0  []        해를 갖거나 갖지 않는 연립일차부등식      NaN      3504                
1  []        절댓값 또는 가우스 기호가 포함된 이차부등식  NaN      3675                
2  []        x축 또는 y축에 접하는 원의 방정식      NaN      3779                
3  []        정수 및 자연수 조건을 갖는 부정방정식     NaN      3649                
4  []        도형의 평행이동 (2)              원        3388                

   question_difficulty  question_type_difficulty  \
0  3                    2.0                        
1  3                    2.0                        
2  3                    2.0                        
3  3                    3.0                        
4  2                    2.0                        

               word2vec_pred_chapter  \
0  ['H1S1-08', 'H1S1-08', 'H1S1-08']   
1  ['H1S1-08', 'HSU1-06', 'H1S1-08']   
2  ['H1S1-11', 'H1S1-12', 'H1S1-11']   
3  ['H1S1-07', 'H1S1-10', 'H1S1-08']   
4  ['H1S1-12', 'H1S1-06', 'H1S2-03']   

                                                           word2vec_pred_name  \
0  ['연립이차부등식', '정수인 해의 개수가 주어진 연립부등식', '해를 갖거나 갖지 않는 연립일차부등식']                  
1  ['이차부등식', '삼각부등식 (2)', '절댓값 또는 가우스 기호가 포함된 이차부등식']                           
2  ['x축 또는 y축에 접하는 원의 방정식', '평행이동과 대칭이동의 활용', 'x축과 y축에 동시에 접하는 원의 방정식']         
3  ['정수 및 자연수 조건을 갖는 부정방정식', '평행한 두 직선 사이의 거리', '이차부등식을 이용하여 식의 최댓값·최솟값 구하기']   
4  ['도형의 평행이동 (2)', '식의 최댓값과 최솟값 (2)', '코시-슈바르츠의 부등식 (1)']                      

  tfidf_pred_chapter    학년 중단원 소단원  
0  H1S1-08            H1S1  08  05  
1  H1S1-08            H1S1  08  11  
2  H1S1-11            H1S1  11  05  
3  H1S1-07            H1S1  07  18  
4  H1S1-01            H1S1  12  03

## 검색하기

### 중단원별 keyword (from soynlp)

In [37]:
# 중단원_summary.csv
df_summary = pd.read_csv("data/중단원_summary.csv", encoding="utf-8")
df_summary

학년  중단원                                         소단원_keywords_soynlp
0   H1S1  1    ['곱셈공식']                                                  
1   H1S1  2    ['나머지정리', '항등식', '계수']                                    
2   H1S1  3    ['인수분해']                                                  
3   H1S1  4    ['복소수', '조건']                                             
4   H1S1  5    ['이차방정식', '이차식', '판별', '조건', '근']                         
5   H1S1  6    ['그래프']                                                   
6   H1S1  7    ['삼차방정식', '사차방정식', '부정방정식', '계수', '조건', '근']              
7   H1S1  8    ['연립일차부등식', '연립이차부등식', '이차부등식', '부등식']                    
8   H1S1  9    ['외분점', '거리', '사이']                                       
9   H1S1  10   ['직선', '사이', '거리', '점', '위']                              
10  H1S1  11   ['방정식', 'x축', '원', '점', '위', '접']                         
11  H1S1  12   ['평행이동', '대칭이동', '점']                                     
12  H1S2  1    ['부분집합', '유한집합', '집합', '원소', '관계']                        
13  H1S2  2    ['성질', '집합', '연산', '개수', '부분', '부분집합']                    
14  H1S2  3    ['절대부등식', '필요조건', '삼단논법', '거짓', '이용', '명제', '대우', '참']    
15  H1S2  4    ['일대일대응', '합성함수', '역함수', 'f∘g', '성질', '함수', '조건']         
16  H1S2  5    ['유리식', '그래프', '차수)', '번분수']                              
17  H1S2  6    ['그래프', '역함수', '위', '값']                                  
18  H1S2  7    ['만족', '이용', '분할', '순열', '조합', '조건', '이웃', '포함', 'C', '수']
19  HSTA  1    ['중복순열']                                                  
20  HSTA  2    ['(a+b)ⁿ', '전개식']                                         
21  HSTA  3    ['덧셈정리', '확률', '기']                                       
22  HSTA  4    ['조건부확률', '곱셈정리', '독립', '종속']                             
23  HSTA  5    ['이산확률변수', '이항분포']                                        
24  HSTA  6    ['정규분포', '관계']                                            
25  HSTA  7    ['모평균']                                                   
26  HSU1  1    ['거듭제곱근', '이용', '지수', '꼴', '값']                           
27  HSU1  2    ['자릿수', '성질', '부분', '로그', '조건', '정의', '밑', '대']           
28  HSU1  3    ['지수방정식', '지수부등식', '지수함수', '그래프', '조건']                   
29  HSU1  4    ['로그방정식', '로그함수', '그래프']                                  
30  HSU1  5    ['삼각함수', '넓이', '호']                                       
31  HSU1  6    ['삼각방정식', '삼각함수', '그래프', '포함', '최소', '활용', '값']           
32  HSU1  7    ['코사인법칙', '사인법칙', '삼각', '넓이']                             
33  HSU1  8    ['등차수열', '일반항', '합', '수']                                 
34  HSU1  9    ['등비수열', '일반항', '합', '수']                                 
35  HSU1  10   ['일반항', '군수열', '포함', '여러', '형태', '분수', '합', '식', '∑']     
36  HSU1  11   ['수열', '정의']

### 중단원/소단원 검색

In [13]:
# 검색어 지정
STR_YEAR_SEMESTER = "HSU1"
STR_MIDDLE_NO = "06"

In [27]:
df_preprocessed[(df_preprocessed['학년']==f'{STR_YEAR_SEMESTER}') & (df_preprocessed['중단원']==f'{STR_MIDDLE_NO}')][['text', 'math_field', 'hangul_field', '학년', '중단원', '소단원']]

text  \
40    삼각형 `ABC`의 세 내각의 크기를 각각 `A`, `B`, `C`라 할 때, `tanA=-1/2`이다. 이때 `sqrt5 cos(B+C)`의 값을 구하시오.                                                                                                                                                                                                                                           
93    `-3<= a<= 3`을 만족하는 정수 `a`의 값들 중 함수 `f(x)=sinx`에 대해 `f(a)`값이 가장 큰 정수 `a`값을 구하시오.                                                                                                                                                                                                                                                    
94    삼각형 `ABC`에서 `4sin^2{::}(B+C)/2=3`이 성립할 때, `cosA`의 값을 구하면?                                                                                                                                                                                                                                                                          
100   함수 `y=sin^2x-2cosx+1`의 최댓값 `a`와 최솟값 `b`의 곱을 구하여라.                                                                                                                                                                                                                                                                                  
142   예각삼각형 `ABC`에서 `2sin^2A-3sinA+1=0`이 성립할 때, `A`값을 구하면?                                                                                                                                                                                                                                                                               
163   함수 `y=|cos{::}x/2|-1`의 주기와 최댓값, 최솟값을 옳게 구한 것을 고르시오.                                                                                                                                                                                                                                                                                
191   `2cos^2theta-3sintheta=3`을 만족하는 모든 `theta`의 합을 `alphapi`라 하자. `10alpha`의 값을 구하시오. <br/>(단, `0<= theta< 2pi`)<br/>                                                                                                                                                                                                                  
234   다음 중 `y=cos{::}x/2`의 그래프를 `x`축의 방향으로 `pi`만큼 평행이동한 후 `y`축에 대하여 대칭이동한 그래프가 나타내는 함수의 식은?                                                                                                                                                                                                                                              
245   함수 `y=(-sinx+1)/(sinx+2)`의 최댓값을 `M`, 최솟값을 `m`이라 할 때, `M+m`의 값은?                                                                                                                                                                                                                                                                    
260   `f(x)=asin{::}x/b+c (a>0, b>0)`의 최댓값이 `5`, 주기가 `4pi`이고, `f(pi/3)=7/2`일 때, `a+b-c`의 값은?                                                                                                                                                                                                                                             
263   두 함수 `y=abs(cosax)`와 `y=abs(tan{::}x/a)`의 주기가 같을 때, 양수 `a`의 값은?                                                                                                                                                                                                                                                                    
288   다음 조건을 만족하는 상수 `a`, `b`, `c`에 대하여 `a+b+c`의 값은? <fieldset><legend></legend>(가) `sin^2 10˚+sin^2 20˚+cdots+sin^2 80˚+sin^2 90˚=a`<br/>(나) `cos10˚+cos20˚+cdots+cos170˚+cos180˚=b`<br/>(다) `tan1˚xxtan2˚xxcdotsxxtan88˚xxtan89˚=c`</fieldset>                                                                                         
298   예각삼각형 `ABC`의 세 내각의 크기를 `A`

### 키워드 검색

In [10]:
keyword = 'log'

# 조건를 충족하는 데이터를 필터링
contains_keyword = df_preprocessed['text'].str.contains(keyword)
subset_df = df_preprocessed[contains_keyword]

print(subset_df.shape)
print(subset_df['chapter'].unique())
subset_df.head()

(293, 16)
['HSU1-04' 'HSU1-02' 'HSU1-03' 'HSU1-09' 'HSTA-03' 'HSU1-10' 'HSU1-08'
 'HSTA-02']


chapter  \
102  HSU1-04   
141  HSU1-04   
168  HSU1-02   
192  HSU1-04   
238  HSU1-02   

                                                                                            text  \
102  연립방정식 `{(log_(2)(x-1)-log_(4)(2y-1)=0, ), (2y-x=1, ):}`의 해를 구하시오.                             
141  연립부등식 `{(2log_2.71(x+1) < log_2.71{::}4(x+1)), (3^(x-1) > 1/9) :}`을 만족시키는 정수 `x`의 개수를 구하시오.   
168  `2^a=3`, `3^b=5`일 때, `log_(60)12`을 `a`, `b`의 식으로 나타내면?                                        
192  연립부등식 `{((1/sqrt2)^(2x-1) < 1/4, ), (x^logx < 1000(1/x)^2, ):}`을 만족하는 정수 `x`의 개수는?            
238  `log_(a-2) (-a^2 + 7a - 10)`가 정의되기 위한 정수 `a`의 값은?                                             

                                                              math_field  \
102  ['{(log_(2)(x-1)-log_(4)(2y-1)=0, ), (2y-x=1, ):}']                   
141  ['{(2log_2.71(x+1) < log_2.71{::}4(x+1)), (3^(x-1) > 1/9) :}', 'x']   
168  ['2^a=3', '3^b=5', 'log_(60)12', 'a', 'b']                            
192  ['{((1/sqrt2)^(2x-1) < 1/4, ), (x^logx < 1000(1/x)^2, ):}', 'x']      
238  ['log_(a-2) (-a^2 + 7a - 10)', 'a']                                   

                              hangul_field  \
102  ['연립방정식', '해', '구하시오']                  
141  ['연립부등식', '만족시키', '정수', '개수', '구하시오']   
168  ['일', '때', '식', '나타내']                  
192  ['연립부등식', '만족하', '정수', '개수']            
238  ['정의되기', '위', '정수', '값']                

                                 tag_field                   name  \
102  []                                     로그방정식 (4)               
141  ['< log_2.71{::}4(x+1)), (3^(x-1) >']  로그부등식 (4)               
168  []                                     로그의 성질의 활용 (2)          
192  []                                     로그부등식 (4)               
238  []                                     로그가 정의되기 위한 밑과 진수의 조건   

            subtitle  qplay_question_id  question_difficulty  \
102  연립방정식으로 제시되는 경우  8955               3                     
141  연립부등식으로 제시되는 경우  7050               3                     
168  aˣ=b 가 주어진 경우    8453               4                     
192  연립부등식으로 제시되는 경우  7702               4                     
238  NaN              8567               3                     

     question_type_difficulty              word2vec_pred_chapter  \
102  3.0                       ['HSU1-04', 'HSU1-04', 'HSU1-04']   
141  3.0                       ['HSU1-04', 'HSU1-04', 'HSU1-04']   
168  2.0                       ['HSU1-02', 'HSU1-02', 'H1S2-03']   
192  3.0                       ['HSU1-04', 'HSU1-03', 'HSU1-03']   
238  1.0                       ['H1S1-02', 'HSU1-02', 'HSU1-04']   

                                                word2vec_pred_name  \
102  ['로그부등식 (4)', '로그방정식 (1)', '로그방정식 (4)']                         
141  ['로그부등식 (4)', '로그부등식 (1)', '로그방정식 (1)']                         
168  ['조건을 이용하여 식의 값 구하기', '로그의 성질의 활용 (2)', '산술평균과 기하평균의 관계 (3)']   
192  ['로그부등식 (4)', '지수부등식 (1)', '지수부등식 (4)']                         
238  ['복잡한 항등식의 계수의 합', '로그의 정의', '로그함수의 최대, 최소 (4)']                

    tfidf_pred_chapter    학년 중단원 소단원  
102  HSU1-04            HSU1  04  14  
141  HSU1-04            HSU1  04  18  
168  HSU1-02            HSU1  02  07  
192  H1S1-01            HSU1  04  18  
238  HSU1-02            HSU1  02  02

In [9]:
search = '다음 중 옳은 것은?'

# 조건를 충족하는 데이터를 필터링
subset_df = df_preprocessed[df_preprocessed['text'] == search]

print(subset_df.shape)
print(subset_df['chapter'].unique())
subset_df.head()

(10, 16)
['HSU1-01' 'H1S1-04' 'HSU1-06' 'H1S2-01' 'H1S2-04' 'H1S1-03']


chapter         text math_field           hangul_field tag_field  \
35    HSU1-01  다음 중 옳은 것은?  []         ['다음', '중', '옳', '것']  []         
47    H1S1-04  다음 중 옳은 것은?  []         ['다음', '중', '옳', '것']  []         
792   HSU1-06  다음 중 옳은 것은?  []         ['다음', '중', '옳', '것']  []         
1324  H1S2-01  다음 중 옳은 것은?  []         ['다음', '중', '옳', '것']  []         
1460  H1S2-04  다음 중 옳은 것은?  []         ['다음', '중', '옳', '것']  []         

                 name subtitle  qplay_question_id  question_difficulty  \
35    거듭제곱근            NaN      5459               3                     
47    복소수의 뜻과 분류       NaN      4371               3                     
792   삼각함수 값의 대소관계 판별  NaN      7055               2                     
1324  유한집합의 원소의 개수     NaN      19511              3                     
1460  항등함수와 상수함수       NaN      24743              2                     

      question_type_difficulty              word2vec_pred_chapter  \
35    1.0                       ['H1S2-03', 'H1S2-01', 'HSU1-05']   
47    1.0                       ['H1S2-03', 'H1S2-01', 'HSU1-05']   
792   1.0                       ['H1S2-03', 'H1S2-01', 'HSU1-05']   
1324 NaN                        ['H1S2-03', 'H1S2-01', 'HSU1-05']   
1460 NaN                        ['H1S2-03', 'H1S2-01', 'HSU1-05']   

                word2vec_pred_name tfidf_pred_chapter    학년 중단원 소단원  
35    ['명제', '집합의 뜻', '호도법과 육십분법']  H1S1-01            HSU1  01  01  
47    ['명제', '집합의 뜻', '호도법과 육십분법']  H1S1-01            H1S1  04  01  
792   ['명제', '집합의 뜻', '호도법과 육십분법']  H1S1-01            HSU1  06  02  
1324  ['명제', '집합의 뜻', '호도법과 육십분법']  H1S1-01            H1S2  01  05  
1460  ['명제', '집합의 뜻', '호도법과 육십분법']  H1S1-01            H1S2  04  08

In [72]:
# Tokenized 결과 확인 from data_josa_removed.csv

df_tokenized = pd.read_csv("data/data_josa_removed.csv", encoding="utf-8")
print(df_tokenized.shape)

QUESTION_ID = 7348

# 조건를 충족하는 데이터를 필터링
df_tokenized[df_tokenized['qplay_question_id']==QUESTION_ID]

(5364, 10)


Unnamed: 0  qplay_question_id  \
260  260         7348                

                                                                                                                                                                                                                                                                                                                      text  \
260  ['`', 'f', '(', 'x', ')', '=', 'a', 'sin', '{', '::', '}', 'x', '/', 'b', '+', 'c', '(', 'a', '>', '0', ',', 'b', '>', '0', ')', '`', '최댓값', '`', '5', '`,', '주기', '`', '4', 'pi', '`', ',', '`', 'f', '(', 'pi', '/', '3', ')', '=', '7', '/', '2', '`', '일', '때', ',', '`', 'a', '+', 'b', '-', 'c', '`', '값', '?']   

           qtid                  name    subtitle  question_difficulty  \
260  HSU1-06-06  삼각함수 식의 미정계수 구하기 (1)  조건이 주어진 경우  3                     

     question_type_difficulty  grade purpose  
260  2.0                       10     open

### 기존

In [ ]:
# Math symbols
math_terms = ['sin', 'cos', 'tan', #Triangle function
             'alpha', 'beta', 'theta', 'gamma', 'omega', 'phi', #Roman letters
             'uu', 'nn', 'sup', 'sub', '^C', #Sets
             'pi', 'abs', 'sqrt', 'cdots', #Etc
             '!', '(', ')', '{', '}', '[', ']', '+', '|', '/_', '@', '=' 
             ]

# 좌표 괄호와 일반 괄호 구분 : 추가 예정

# Math symbols which need regular expression
math_terms_complex = [r'_[0-9]+C_[0-9+]', #Combination
                     r'_[0-9]+P_[0-9+]' #Permutation
                     ] 

# Symbols which are not related to math
non_math_terms = ['<br/>', '<fieldset>', '<legend>', '</fieldset>', '</legend>', '\{::\}']

### 의논사항

In [ ]:
'sqrt', 'sin', 'cos', 'tan', '{::}', '|', 'theta', 'alpha', 'beta', 'gamma','delta' ,'pi', 'abs', 'cdots', 'sin^2', 'cos^2', '˚', '/_', '△', 'lim_'
'n->', 'oo', 'sum_', 'bar'

- `{::}`는 함수식 내 분수를 표현하는 기호. 없애는 것보다 따로 처리해주면 성능이 올라갈까?
- `sin^`, `cos^`, `tan^` 를 따로 분류하면? 
- 곱하기 기호가 `x` 문자로 표현되는 경우가 있는 것 같다. (tan1˚xxtan2˚xxcdotsxxtan88˚xxtan89˚=c, Mxxm 등) -> `xx`로 분리되고 있음
- A_1 P_(n-1) 아래첨자 표기법도 Combination, Permutation 처럼 하나로 인식하면 좋을 것 같다. (A, _, 1 로 분리되어 인식되고 이 ㅆ음)
- `|` 는 절댓값일수도, 조건 bar 일수도(정의역, 치역) -> 물론, '정의역', '치역'이라는 단어로 분리되긴 할 듯.
- `oo`  infinity 기호 -> `oo`로 잘 인식하고 있음
- `sum`도 math_terms에 -> ex) `lim_(n->oo)1/nsum_(k=1)^(n-1)1/S_k` ->  'lim', '_', '(', 'n', '->', 'oo', ')', '1', '/', 'ns', 'um', '_', '(', 'k', '=', '1', ')', '^', '(', 'n', '-1', ')', '1', '/', 'S', '_', 'k' 로 인식되고 있음. `ns와 um으로 분리되어 인식되고 있음`
- `squareABCD`가 하나로 인식되고 있음(data 개수 2개) -> 큰 상관은 없을 듯하다
- 의미중복 ABCD: 사각형? 정사면체? / () : 점? 좌표? / | : 정의역치역? 절댓값?

- 

- cosax 제대로 나뉘는지
- {::} 는 함수식 내 분수에서 쓰이는 것 같다
- sin^2x를 잘 구분하는가
- fieldset, legend, br, 보기 는 한번에 인식해야지
- cdots 는 ...
- tan1˚xxtan2˚xxcdotsxxtan88˚xxtan89˚=c 인식할까? (288) -> 곱하기가 x로 나오나봐. (911 Mxxm)
- A_1 P_(n-1) 아래첨자 표기법 -> 한 번에 인식하면 좋을 텐데
- '다음 중 옳은 것은'은 분류가 되나? 4개 캡처 (7411번 다음 함수 중 주기가 가장 큰 것은? 삼각함수를 포함한 식의 최대, 최소와 주기, 8111(2816)번 주기가 다른 하나 절댓값 기호가 포함된 삼각함수   )
- | 는 절댓값일수도, 조건 bar 일수도(정의역, 치역)
- Api, Bpi 잘 분리?
- 호나 선분 AB 이런건 분리 안되면 좋긴 한디
- 'oo' 은 무한대야
- lim_(n->oo)1/nsum_(k=1)^(n-1)1/S_k
- 점(좌표)는 다르게 취급?
- cospix 이런식으로 pi의 위치 찾아보기 삼각함수말고 pi와 결합
- bar 선분표기
- squareABCD 붙어있는 경우